Transaction flex trit

ifullgaz opened the pull request:
Rewrite of transaction to support flex trit

ifullgaz committed:
entangled/transaction: rewrite to support flex trits

ifullgaz committed:
entangled/transaction: rewrite to support flex trits with bounds checking

th0br0 commented:
@ifullgaz could you rebase on master please

ifullgaz committed:
entangled/model: added build targets with default 1:1 memory model

th0br0 commented on common/model/BUILD:
```diff
@@ -1,10 +1,41 @@
+    name="transaction",
+    srcs=["transaction.c"],
+    hdrs=["transaction.h"],
+    defines = ["TRIT_ARRAY_ENCODING_1_TRIT_PER_BYTE"],
```
don't add the `defines` here

th0br0 commented on common/model/BUILD:
```diff
@@ -1,10 +1,41 @@
+    name="transaction_1",
+    srcs=["transaction.c"],
+    hdrs=["transaction.h"],
+    defines = ["TRIT_ARRAY_ENCODING_1_TRIT_PER_BYTE"],
```
don't add the `defines` here, it's already provided by `trit_array_1`

th0br0 commented on common/model/BUILD:
```diff
@@ -1,10 +1,41 @@
+    name="transaction_4",
+    srcs=["transaction.c"],
+    hdrs=["transaction.h"],
+    defines = ["TRIT_ARRAY_ENCODING_4_TRITS_PER_BYTE"],
```
same here

th0br0 commented on common/model/BUILD:
```diff
@@ -1,10 +1,41 @@
+    name="transaction_5",
+    srcs=["transaction.c"],
+    hdrs=["transaction.h"],
+    defines = ["TRIT_ARRAY_ENCODING_5_TRITS_PER_BYTE"],
```
and here

th0br0 commented on common/model/tests/BUILD:
```diff
@@ -1,8 +1,32 @@
+cc_test (
+    name = "test_transaction_1",
     srcs = ["test_transaction.c"],
+    defines = ["TRIT_ARRAY_ENCODING_1_TRIT_PER_BYTE"],
```
no need

th0br0 commented on common/trinary/tests/BUILD:
```diff
@@ -55,7 +55,7 @@ cc_test(
     name = "test_trit_array_1",
     srcs = ["test_trit_array.c"],
-    copts = ["-DTRIT_ARRAY_ENCODING_1_TRIT_PER_BYTE"],
+    defines = ["TRIT_ARRAY_ENCODING_1_TRIT_PER_BYTE"],
```
remove this line

th0br0 reviewed (changes requested):
could you run `clang-format` on the files please?

ifullgaz committed:
entangled/model: clang-format

ifullgaz committed:
entangled/trinary: added conversion function from/to trytes to/from flex_trits
entangled/model: added cpp wrapper class around iota_transaction structure
entangled/cppclient: rewrite to use cpptransaction and flexible trits

ifullgaz committed:
entangled/model: added missing test file

ifullgaz committed:
entangled/trinary: fix memory not initialized

th0br0 commented on common/model/cpptransaction.cc:
```diff
@@ -0,0 +1,241 @@
+#include "common/model/cpptransaction.h"
+#include "common/trinary/trit_array.h"
+
+namespace model {
```
let's namespace this `iota::model::Transaction`
(also, classes should start with an upper case letter)

th0br0 commented on common/model/BUILD:
```diff
@@ -1,10 +1,77 @@
+        "//common/trinary:trit_array"
+    ], )
+
+cc_library(
```
There's no need for the various `_{1,4,5}` targets. We're already testing `transaction_{1,4,5}` elsewhere. So one target should be sufficient

th0br0 commented on common/model/BUILD:
```diff
@@ -1,10 +1,77 @@
+    ], )
+
+cc_library(
+    name="transaction_4",
```
same goes for this

th0br0 commented on common/model/cpptransaction.cc:
```diff
@@ -0,0 +1,241 @@
+}
+
+std::string transaction::get_signature(void) {
+  auto trits = transaction_signature(iota_transaction);
```
assuming that we want to return a `string` here (which I'm not so sure of), you should much rather first create a fixed-length string and then write directly to it instead of copying twice

th0br0 commented on common/model/cpptransaction.cc:
```diff
@@ -0,0 +1,241 @@
+
+// Get the transaction message
+std::string transaction::get_message(void) {
+  auto trits = transaction_message(iota_transaction);
```
same here (as `get_signature`)

th0br0 replied:
also, shouldn't this call `get_signature`? why is this method even needed?

th0br0 commented on common/model/tests/BUILD:
```diff
@@ -1,8 +1,65 @@
+cc_test (
+    name = "test_transaction_4",
+    srcs = ["test_transaction.c"],
+    defines = ["TRIT_ARRAY_ENCODING_4_TRITS_PER_BYTE"],
```
remove

th0br0 commented on common/model/tests/BUILD:
```diff
@@ -1,8 +1,65 @@
+)
+
+cc_test (
+    name = "test_transaction_4",
```
we don't really need to test each encoding on a transaction level as well

th0br0 replied:
this also applies to the targets further below

th0br0 commented on common/model/tests/test_cpptransaction.cc:
```diff
@@ -0,0 +1,16 @@
+ * Refer to the LICENSE file for licensing information
+ */
+
+#include <string.h>
```
Either write tests (which you should) or remove this file.
Also, C++ tests use googletest

th0br0 commented on common/model/transaction.c:
```diff
@@ -21,97 +21,146 @@
-  transaction_set_address(transaction, trytes + offset);
-  offset += sizeof(transaction->address);
-  transaction_set_value(transaction, trytes_to_long(trytes + offset, 27));
+size_t transaction_deserialize_trits(iota_transaction_t transaction,
```
should rather be `deserialize_flex_trits`?

th0br0 reviewed (changes requested):
Could you create a C++ wrapper for `trit_array_p` and just use that on the C++ Transaction side as well?

jserv commented on common/model/cpptransaction.cc:
```diff
@@ -0,0 +1,241 @@
+/*
+ * Copyright (c) 2018 IOTA Stiftung
```
Why didn't copyright owner belong to IOTA Foundation?

th0br0 replied:
The correct legal name is `IOTA Stiftung`. `IOTA Foundation` is just the english translation of that. Both work arguably.

jserv commented on common/model/cpptransaction.h:
```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright (c) 2018 IOTA Stiftung
```
Ditto.

jserv commented on common/model/tests/test_cpptransaction.cc:
```diff
@@ -0,0 +1,16 @@
+/*
+ * Copyright (c) 2018 IOTA Stiftung
```
Ditto.

jserv commented on common/model/transaction.h:
```diff
@@ -1,6 +1,6 @@
 /*
  * Copyright (c) 2018 IOTA Stiftung
```
Ditto.

jserv reviewed (commented):
Clarify copyright owner

ifullgaz committed:
entangled/model: added namespace iota, capital letter for CPP Transaction class

ifullgaz committed:
entangled/trinary: cpp wrapper for flex_trit

ifullgaz committed:
entangled/trinary: cpp wrapper for flex_trit

ifullgaz committed:
entangled/cppclient: use Transaction class

ifullgaz committed:
entangled: applied clang_format

ifullgaz committed:
entangled/trinary: added missing include for cassert

ifullgaz committed:
entangled/trinary: optimisations to FlexTritArray

tsvisabo commented on common/trinary/flex_trit_array.h:
```diff
@@ -0,0 +1,102 @@
+    int index;
+
+   public:
+    FlexTritArrayProxy(FlexTritArray &flex_trit_array, int index)
```
add const

tsvisabo commented on common/trinary/flex_trit_array.h:
```diff
@@ -0,0 +1,102 @@
+  /// current memory model.
+  /// @param[in] num_trits - number of trits to store
+  /// @return size_t - the number of bytes need
+  static size_t sizeForTrits(size_t num_trits) {return flex_trits_num_for_trits(num_trits);};
```
I think numBytesForTrits is clearer

tsvisabo commented on common/trinary/flex_trit_array.h:
```diff
@@ -0,0 +1,102 @@
+  /// Returns a new FlexTritArray from a vector of trits.
+  /// @param[in] trits - a vector of trits
+  /// @return FlexTritArray
+  static FlexTritArray createFromTrits(const std::vector<trit_t> trits);
```
Pass by &

tsvisabo replied:
You should define a CTOR rather than a static method

tsvisabo commented on common/trinary/flex_trit_array.h:
```diff
@@ -0,0 +1,102 @@
+  /// Returns a new FlexTritArray from a vector of trytes.
+  /// @param[in] trytes - a vector of trytes
+  /// @return FlexTritArray
+  static FlexTritArray createFromTrytes(const std::vector<tryte_t> trytes);
```
Pass by &

tsvisabo commented on common/trinary/flex_trit_array.h:
```diff
@@ -0,0 +1,102 @@
+  /// @param[in] flex_trits - a vector of flex_trits
+  /// @param[in] num_trits - the number of trits encoded in the vector of flex_trits
+  /// @return FlexTritArray
+  FlexTritArray(const std::vector<flex_trit_t> flex_trits, size_t num_trits) : flex_trits_(flex_trits), num_trits_(num_trits) {};
```
Pass by &

tsvisabo commented on common/trinary/flex_trit_array.h:
```diff
@@ -0,0 +1,102 @@
+  /// Destructor
+  ~FlexTritArray() {};
+
+  FlexTritArrayProxy operator[](const int i) { return FlexTritArrayProxy(*this, i); }
```
no need for const, it's a copied arg

tsvisabo commented on common/trinary/flex_trit_array.cc:
```diff
@@ -0,0 +1,61 @@
+
+namespace iota {
+namespace trinary {
+  FlexTritArray FlexTritArray::createFromTrits(const std::vector<trit_t> trits) {
```
Pass &

tsvisabo commented on common/trinary/flex_trit_array.cc:
```diff
@@ -0,0 +1,61 @@
+namespace trinary {
+  FlexTritArray FlexTritArray::createFromTrits(const std::vector<trit_t> trits) {
+    size_t flex_len = FlexTritArray::sizeForTrits(trits.size());
+    std::vector<flex_trit_t> flex_trits;
```
you don't have to resize, just pass the size to the CTOR

tsvisabo commented on common/trinary/flex_trit_array.h:
```diff
@@ -0,0 +1,102 @@
+  FlexTritArray(const FlexTritArray &other) : flex_trits_(other.flex_trits_), num_trits_(other.num_trits_) {};
+
+  /// Destructor
+  ~FlexTritArray() {};
```
no need

ifullgaz committed:
entangled/trinary: optimisations to FlexTritArray

th0br0 commented on common/model/cpptransaction.h:
```diff
@@ -0,0 +1,96 @@
+
+ public:
+  Transaction();
+  Transaction(std::string trytes);
```
`const std::string& trytes`

th0br0 commented on common/model/cpptransaction.h:
```diff
@@ -0,0 +1,96 @@
+
+  std::string get_signature(void);
+  // Set the transaction signature (copy argument)
+  void set_signature(std::string signature);
```
`const std::string&`

th0br0 commented on common/model/cpptransaction.h:
```diff
@@ -0,0 +1,96 @@
+  // Get the transaction message
+  std::string get_message(void);
+  // Set the transaction message (copy argument)
+  void set_message(std::string message);
```
`const std::string&`

th0br0 commented on common/model/cpptransaction.h:
```diff
@@ -0,0 +1,96 @@
+  // Get the transaction address
+  std::string get_address(void);
+  // Set the transaction address (copy argument)
+  void set_address(std::string address);
```
`const std::string&`

th0br0 commented on common/model/cpptransaction.h:
```diff
@@ -0,0 +1,96 @@
+  // Get the transaction value
+  int64_t get_value(void);
+  // Set the transaction value
+  void set_value(int64_t value);
```
`const std::string&`

th0br0 commented on common/model/cpptransaction.cc:
```diff
@@ -0,0 +1,240 @@
+void Transaction::set_address(std::string trytes) {
+  size_t flex_len = flex_trits_num_for_trits(243);
+  flex_trit_t trits[flex_len];
+  tryte_to_flex_trit(trits, 243, (const tryte_t *)trytes.data(), 81, 81);
```
Instead of allocating `flex_trit_t` on the stack, we should write straight to the element on the `iota_transaction` instance. This saves quite a bit of memory!

th0br0 replied:
Same goes for all the other methods here as well

th0br0 commented on common/model/tests/BUILD:
```diff
@@ -1,8 +1,65 @@
+cc_test (
+    name = "test_transaction_5",
+    srcs = ["test_transaction.c"],
+    defines = ["TRIT_ARRAY_ENCODING_5_TRITS_PER_BYTE"],
```
`remove

th0br0 commented on common/model/tests/BUILD:
```diff
@@ -1,8 +1,65 @@
+cc_test (
+    name = "test_cpptransaction_1",
+    srcs = ["test_cpptransaction.cc"],
+    defines = ["TRIT_ARRAY_ENCODING_1_TRIT_PER_BYTE"],
```
`remove

th0br0 commented on common/model/tests/BUILD:
```diff
@@ -1,8 +1,65 @@
+cc_test (
+    name = "test_cpptransaction_4",
+    srcs = ["test_cpptransaction.cc"],
+    defines = ["TRIT_ARRAY_ENCODING_4_TRITS_PER_BYTE"],
```
remove

th0br0 commented on common/model/tests/BUILD:
```diff
@@ -1,8 +1,65 @@
+cc_test (
+    name = "test_cpptransaction_5",
+    srcs = ["test_cpptransaction.cc"],
+    defines = ["TRIT_ARRAY_ENCODING_5_TRITS_PER_BYTE"],
```
remove

th0br0 commented on common/model/tests/test_cpptransaction.cc:
```diff
@@ -0,0 +1,16 @@
+#include <unity/unity.h>
+
+int main(void) {
+  UNITY_BEGIN();
```
Test with GoogleTest

th0br0 commented on common/trinary/flex_trit_array.h:
```diff
@@ -0,0 +1,116 @@
+  /// Returns a new FlexTritArray from a vector of trits.
+  /// @param[in] trits - a vector of trits
+  /// @return FlexTritArray
+  static FlexTritArray createFromTrits(const std::vector<trit_t> &trits);
```
also run `clang-format` on the files?

th0br0 commented on common/trinary/flex_trit_array.h:
```diff
@@ -0,0 +1,116 @@
+  /// starting at index start.
+  /// @param[in] start - the start index in the receiver
+  /// @return size_t - the number of inserted trits.
+  size_t insert(FlexTritArray &flex_trit_array, size_t start);
```
add `const`

ifullgaz committed:
entangled/trinary: bug fixes, added trit access to trytes, cleanup BUILD files
entangled/model: added serialize to cpptransaction, added tests for cpptransaction, cleanup BUILD files

ifullgaz committed:
entangled/client: fix BUILD file

ifullgaz committed:
entangled/trinary: Added 3 trits per byte storage option, added trit_array insert

ifullgaz committed:
entangled/trinary: Added 3:1 tests, fixed transaction serialize for 3:1 storage scheme

ifullgaz committed:
entangled/model: Disabled 3:1 tests

tsvisabo commented on common/model/cpptransaction.cc:
```diff
@@ -64,7 +64,7 @@ std::string Transaction::get_address(void) {
+void Transaction::set_address(const std::string &trytes) {
   size_t flex_len = flex_trits_num_for_trits(243);
   flex_trit_t trits[flex_len];
   tryte_to_flex_trit(trits, 243, (const tryte_t *)trytes.data(), 81, 81);
```
Can you #define/use defined  constants for 243/81/81?

tsvisabo commented on common/trinary/flex_trit_array.h:
```diff
@@ -0,0 +1,102 @@
+    operator trit_t() const {
+      return flex_trit_array_at(flex_trit_array.flex_trits_.data(), flex_trit_array.num_trits_, index);
+    }
+    FlexTritArrayProxy operator=(trit_t t) {
```
Not a huge performance increase, but should return `FlexTritArrayProxy&`

tsvisabo commented on common/trinary/trit_tryte.c:
```diff
@@ -18,6 +18,42 @@ static const trit_t TRYTE_TO_TRITS_MAPPINGS[TRYTE_SPACE]
+}
+
+trit_t get_trit_at(tryte_t *const trytes, size_t const length, size_t index) {
+  size_t tindex = index / 3U;
```
move:
`+  size_t tindex = index / 3U;
+  if (tindex >= length) {
+    return 0;
+  }
+  tryte_t tryte = trytes[tindex];
+  size_t cindex = tryte == '9' ? 0 : tryte - '@';`
to a static function?

ifullgaz committed:
entangled/model: Added #define for trit sizes

th0br0 commented on common/model/BUILD:
```diff
@@ -1,10 +1,83 @@
+    name="transaction_3",
+    srcs=["transaction.c"],
+    hdrs=["transaction.h"],
+    defines = ["TRIT_ARRAY_ENCODING_3_TRITS_PER_BYTE"],
```
`trit_array` already has this `defines`

th0br0 commented on common/model/cpptransaction.cc:
```diff
@@ -0,0 +1,247 @@
+// Get the transaction message
+std::string Transaction::get_message(void) {
+  auto trits = transaction_message(iota_transaction);
+  tryte_t trytes[2187];
```
as `tryte_t` == `char`, we can avoid the copy by accessing the string's `data()` array directly: https://en.cppreference.com/w/cpp/string/basic_string/data

th0br0 replied:
as `tryte_t` == `char`, we can avoid the copy by accessing the string's `data()` array directly: https://en.cppreference.com/w/cpp/string/basic_string/data

th0br0 commented on common/model/cpptransaction.cc:
```diff
@@ -0,0 +1,247 @@
+std::string Transaction::get_signature(void) {
+  auto trits = transaction_signature(iota_transaction);
+  tryte_t trytes[2187];
+  flex_trit_to_tryte(trytes, 2187, trits, 6561, 6561);
```
as `tryte_t` == `char`, we can avoid the copy by accessing the string's `data()` array directly: https://en.cppreference.com/w/cpp/string/basic_string/data

th0br0 commented on common/model/cpptransaction.cc:
```diff
@@ -0,0 +1,247 @@
+// Get the transaction address
+std::string Transaction::get_address(void) {
+  auto trits = transaction_address(iota_transaction);
+  tryte_t trytes[81];
```
as `tryte_t` == `char`, we can avoid the copy by accessing the string's `data()` array directly: https://en.cppreference.com/w/cpp/string/basic_string/data

th0br0 commented on common/model/cpptransaction.cc:
```diff
@@ -0,0 +1,247 @@
+std::string Transaction::get_obsolete_tag(void) {
+  auto trits = transaction_obsolete_tag(iota_transaction);
+  tryte_t trytes[81];
+  flex_trit_to_tryte(trytes, 81, trits, 243, 243);
```
as `tryte_t` == `char`, we can avoid the copy by accessing the string's `data()` array directly: https://en.cppreference.com/w/cpp/string/basic_string/data

th0br0 commented on common/model/cpptransaction.h:
```diff
@@ -0,0 +1,97 @@
+
+  std::string get_signature(void);
+  // Set the transaction signature (copy argument)
+  void set_signature(const std::string& signature);
```
why do these take `std::string` and not `trit_array`?

th0br0 commented on common/model/cpptransaction.cc:
```diff
@@ -0,0 +1,247 @@
+}
+
+// Set the transaction signature (copy argument)
+void Transaction::set_signature(const std::string &trytes) {
```
also, could we `CamelCase` these methods?

th0br0 commented on common/model/cpptransaction.h:
```diff
@@ -0,0 +1,97 @@
+  Transaction(const std::string& trytes);
+  ~Transaction(void);
+
+  std::string get_signature(void);
```
no need to prefix with `get_`? (for this & the other methods)

ifullgaz committed:
entangled/model: Renamed cpptransaction to TryteTransaction, cleanup BUILD files,

th0br0 commented on common/trinary/trit_tryte.c:
```diff
@@ -18,6 +18,42 @@ static const trit_t TRYTE_TO_TRITS_MAPPINGS[TRYTE_SPACE]
+  if (tindex >= length) {
+    return 0;
+  }
+  tryte_t tryte = trytes[tindex];
```
`if(tryte == '9') return 0;`

th0br0 commented on common/trinary/trit_tryte.c:
```diff
@@ -18,6 +18,42 @@ static const trit_t TRYTE_TO_TRITS_MAPPINGS[TRYTE_SPACE]
+         NUMBER_OF_TRITS_IN_A_TRYTE;
+}
+
+trit_t get_trit_at(tryte_t *const trytes, size_t const length, size_t index) {
```
could we call this `trytes_trit_at` or `trytes_get_trit_at`?

th0br0 commented on common/trinary/trit_tryte.c:
```diff
@@ -18,6 +18,42 @@ static const trit_t TRYTE_TO_TRITS_MAPPINGS[TRYTE_SPACE]
+  return TRYTE_TO_TRITS_MAPPINGS[cindex][tindex];
+}
+
+uint8_t set_trit_at(tryte_t *const trytes, size_t const length, size_t index,
```
also fix name here

th0br0 commented on common/trinary/trit_tryte.c:
```diff
@@ -18,6 +18,42 @@ static const trit_t TRYTE_TO_TRITS_MAPPINGS[TRYTE_SPACE]
+    tryte += TRYTE_SPACE;
+  }
+  trytes[tindex] = TRYTE_STRING[(size_t)tryte];
+  return 1;
```
this should return `1` if it failed, not if it succeeded?

th0br0 commented on common/model/tests/BUILD:
```diff
@@ -1,8 +1,79 @@
+        ]
+)
+
+#cc_test (
```
these should be run.

th0br0 commented on common/model/tests/BUILD:
```diff
@@ -1,8 +1,79 @@
+        ]
+)
+
+#cc_test (
```
these should be run?

ifullgaz committed:
entangled/model: Enabled 3:1 tests after bug fix

th0br0 committed:
Merge branch 'develop' into transaction_flex_trit
